[Blackjack - step1] Submit knopgm's mission

knopgm opened the pull request:
# Blackjack Project -Step1

## Overview:
This project aim to mimic the basic rules of a Blackjack game.
It's following the rules of OOP and a approach towards RDD.

## Main Features:
- Classes: [Participant, Player, Dealer, Card, Deck, Hand]
- Views: [ InputView, OutputView, InputValidator ]
- Controllers: [ BlackJack, GameLogic]
- Utils: [ CardSuit, ResultTypes ]

`BlackJack` is the main connection with all functionalities.

`Participant` is an abstract function that will extends as `Player` and `Dealer`.

`Card` defines what one card is and Deck define what list of cards should be created to be valid as a game deck.

`Hand` is the one who knows what cards a hand has and its score.

`GameLogic` was used to build the steps the game needs.

### Current Limitations:
I still have to do Tests for many Functionalities. (Schedule-time)
I know my conditional functions didn't use the rules of Object Calisthenics, and I'll try to improve it in the review code time. (Schedule-time)
I loved to understand and use the abstract class. 
While I was coding huge interactions with a list of players or participants or their related list outputs, I had a tip to interact with only one item of the list at a time, which reduce a lot the use of conditionals and the size of my methods. But I found it hard to decide which one of the involved classes were responsible for some behaviors. -> Like for example in the method `checkWinOrLose`: Who is responsible to check it? The dealer, the participant or the hand? 
As a beginner, It still blurry how to build a code that is clear for any other developer to read and follow. I think architecture will make a difference to a easier understand. Does it? and how much it do make a difference?
_________________
Thank you for your time and thoughtful guidance in advance! 🙏

knopgm committed:
docs(Readme): plan features for a blackjack game

knopgm committed:
docs(Readme): fix readme style

knopgm committed:
feat(Card, CardGenerator and CardSuit): Create features to build a deck of cards

knopgm committed:
feat(Dealer, DealerTest): create dealer class and test firstTurnCards method

knopgm committed:
feat(Participant, Player, Dealer, Card, Deck, DeckGenerator, GameLogic): Create basic classes for a Blackjack game and update Readme file

knopgm committed:
refactor(DeckGenerator): move deck generator logic to be a companion object of Deck class

knopgm committed:
feat (Dealer, Deck): create a method inside Dealer to draw a card from the Deck and test it

knopgm committed:
feat (Hand): create a Hand class, first turn method and save 2 cards for each participant

knopgm committed:
feat (Hand): create method to sum card values and a test for it

knopgm committed:
feat (GameLogic): create a method to map players score

knopgm committed:
feat (GameLogic): create a method to identify the result of each score and other to identify a blackjack to end the game

knopgm committed:
feat (Results): add an enum class to save the result type of the game

knopgm committed:
refactor (Hand, GameLogic): move score methods into Hand`s class responsibility
refactor (Hand): add condition to sum card values when it contains a value in position 1.

knopgm committed:
doc (Readme): update readme check task for calculate participant score with Ace logic

knopgm committed:
refactor (Hand): update sum os card values with rules of j, q and k card values as 10
doc (Readme): refactor readme to better reflect implemented game features

knopgm committed:
refactor (Hand, GameLogic, Participant, Dealer, Player): move score logic to each participant

knopgm committed:
refactor (GameLogic): move results checks to each participant
feat (GameLogic): create a method to run another turn of the game

knopgm committed:
doc (Readme): update checklist of implemented features and refactor the doc to better reflect the game logic

knopgm committed:
feat (Card): create a method to convert card numbers into a card-face element and override the method toString to print a card as a string

knopgm committed:
feat (Hand): override the method toString to print a hand of cards as a string

knopgm committed:
feat (InputView, InputValidator, OutputView): create InputView, and it's InputValidator classes to read user's prompts and a OutputView class to print game logic results

knopgm committed:
fix (GameLogic): use while to ask players how many cards they want

knopgm committed:
feat (OutputView): create methods to print current cards of players after each hit

knopgm committed:
refactor (BlackjackController): connect input and output view classes to the game controller

knopgm committed:
style (GameLogic, Player): fix style of code and add an empty line to the console

knopgm committed:
feat (BlackjackController, Dealer, Hand and OutputView): create print methods to display scores and final results

Wordbe commented on README.md:
```diff
@@ -1 +1,34 @@
+### Hand
+- [x] Calculate participant score with Ace logic
+- [x] Calculate participant score with J, Q and K logic
+- [ ] provide information for game results announce
```
Please check this if you comleted!

Wordbe commented on README.md:
```diff
@@ -1 +1,34 @@
+- [ ] provide information for game results announce
+
+### Game
+- [ ] Accept player names
```
Same here!

Wordbe commented on src/main/kotlin/blackjack/BlackjackController.kt:
```diff
@@ -0,0 +1,31 @@
+        output.printFinalScores(dealer)
+        players.forEach { player -> output.printFinalScores(player) }
+
+        println("## Final Results ##")
```
Displaying output is the responsibility of the View, so let the View handle it! 🙂

Wordbe commented on src/main/kotlin/blackjack/InputView.kt:
```diff
@@ -0,0 +1,19 @@
+
+class InputView {
+    private val validator = InputValidator()
+    val output = OutputView()
```
Try to use OutputView only within the controller.
It's fine for InputView to use println to guide user input! 🙂

Wordbe commented on src/main/kotlin/blackjack/GameLogic.kt:
```diff
@@ -0,0 +1,35 @@
+package blackjack
+
+class GameLogic {
+    fun namesToPlayers(names: List<String>): List<Player> {
```
You can simplify this into a one-liner by removing the variable!

Wordbe commented on src/main/kotlin/blackjack/GameLogic.kt:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    fun otherTurnCards(players: List<Player>, dealer: Dealer) {
+        val output = OutputView()
```
How about making the interaction between the View and the domain model occur only in the Controller? 🤔

knopgm replied:
I can see that my code was violation the responsibility rules of the logic, so I refactor a lot of things about the input/output view (again) in the same commit.
I hope now all the responsibilities are more clear 😅

Wordbe replied:
Great. Thank you!

Wordbe commented on src/main/kotlin/blackjack/Hand.kt:
```diff
@@ -0,0 +1,34 @@
+   fun sumCards(): Int {
+        val cardValues = cards.map { card -> card.number }.toMutableList()
+        val elevenSumCheck = cardValues.sum() + 10
+        cardValues.replaceAll { if (it == 1 && elevenSumCheck <= 21) 11 else it }
```
It would be good to define magic numbers as constants! 
Aside from what's inside this class, please change everything else that’s visible!

Wordbe commented on src/main/kotlin/blackjack/Hand.kt:
```diff
@@ -0,0 +1,34 @@
+        }
+    }
+
+    override fun toString(): String {
```
If you create elements needed for the view inside a domain object, it introduces a dependency on the view. This means that when the view changes, the domain object must also be modified. Try moving this logic into the view instead! 🙂

Wordbe commented on src/main/kotlin/blackjack/Dealer.kt:
```diff
@@ -0,0 +1,18 @@
+        val scoreHand = this.sumCards()
+        if (scoreHand < 17) {
+            println()
+            println("Dealer draws one more card due to having 16 or less.")
```
Instead of having OutputView handle this part, try throwing an exception here. 😊

Wordbe commented on src/main/kotlin/blackjack/Deck.kt:
```diff
@@ -0,0 +1,24 @@
+package blackjack
+
+class Deck(cards: List<Card>) {
+    val cards = cards.shuffled().toMutableList()
```
You can declare properties together in the primary constructor. Try combining them.

Also, functions like shuffled() that return random results make testing difficult. How about moving that logic outside of Deck so it’s easier to test? 😊

knopgm replied:
I created a boolean parameter for the deck generator function, so I can use the function with false to test it, and the function with true to the game logic 😅  I hope that fits the solution you are looking for. 🙏

Wordbe replied:
What I meant wasn’t to add a new field, but rather to move out behavior like `shuffled()`—which has an uncontrollable return value—from the object and pass it in as a parameter instead. That way, your code will be much easier to test! 😊

Wordbe commented on src/main/kotlin/blackjack/Hand.kt:
```diff
@@ -0,0 +1,34 @@
+package blackjack
+
+class Hand() {
+    val cards: MutableList<Card> = mutableListOf()
```
Try creating it in the primary constructor—it improves readability and makes testing easier.

Wordbe commented on src/main/kotlin/blackjack/InputValidator.kt:
```diff
@@ -0,0 +1,10 @@
+    fun validateNames (input: List<String>) {
+        input.forEach { name ->
+            require(name.isNotBlank()) { "Input cannot be blank." }
+            require(name.chars().count() > 2) { "Name should have more than 2 characters." }
```
There should be a function to get the length of a string. Try using it!

Wordbe commented on src/main/kotlin/blackjack/Participant.kt:
```diff
@@ -0,0 +1,27 @@
+
+    abstract fun shouldHit(): Boolean
+
+
```
Please remove the blank spaces!

Wordbe commented on src/main/kotlin/blackjack/Player.kt:
```diff
@@ -0,0 +1,16 @@
+        if (scoreHand < 21) {
+            println()
+            println("Would ${this.name} like to draw another card? (y for yes, n for no)")
+            val wantHit = readln().trim()
```
The InputView is performing tasks that should be handled within the domain object. It looks like a change is needed. 
(println, readln)

Wordbe commented on src/main/kotlin/blackjack/ResultTypes.kt:
```diff
@@ -0,0 +1,7 @@
+package blackjack
+
+enum class ResultTypes() {
```
You can omit the default constructor since it’s automatically generated. 🙂

Wordbe commented on src/test/kotlin/blackjack/DealerTest.kt:
```diff
@@ -0,0 +1,15 @@
+        Dealer(deck).drawCard()
+        val deckSize = deck.cards.size
+
+        assertEquals(47, deckSize, "Should draw a card out of the deck")
```
Don't just test the count—create specific cards and check if they match as expected.
If you extract the random generation part, it becomes testable.

knopgm replied:
I created a test with specific cards to see if the right card is draw. :)
I'll try to implement more test later.

Wordbe reviewed (changes requested):
Hello Gabi!

Thank you for diligently working on the Blackjack mission and submitting your PR. 🙌

I've left comments on areas that might be worth reconsidering. Please take a look and reflect any changes you find appropriate!

Also, the tests could use some reinforcement—I'd appreciate it if you could enhance them.

Thanks again! 😊

knopgm committed:
refactor (BlackJackController and OutputView): move print outputs to the output view file

knopgm committed:
refactor (BlackJackController and InputView): move outputs print call from input view to the controller

knopgm committed:
refactor (GameLogic): reduce function size by returning directly the mapping result

knopgm committed:
refactor(GameLogic): decouple input/output from game logic
- Removed view responsibilities (InputView/OutputView) from GameLogic
- Moved user interaction logic to BlackjackController
- GameLogic now only handles pure decision-making and game rules
- Improves separation of concerns and testability

knopgm committed:
refactor(Hand): implement magic numbers

knopgm committed:
fix (Hand): fix the logic of result winners by comparing players result with dealer result

knopgm committed:
refactor (Hand, OutputView): move format function logic to the outputview responsibility

knopgm committed:
refactor (Hand): move cards property initialization into the primary constructor for readability, improvement of testability and more idiomatic

knopgm committed:
refactor (InputValidator): improve validation by the use of length to check the amount of characters in a name input

knopgm committed:
chore (Participant): remove blank spaces and unused function

knopgm committed:
chore (ResultTypes): remove unused enum

knopgm committed:
refactor (Deck): add a boolean parameter to the generate deck function where the function shuffles the cards if needed

knopgm committed:
chore (OutputView): remove unused expression

knopgm committed:
test (DealerTest): implement test to draw the correct card

knopgm commented:
Hy [Wordbe](https://github.com/Wordbe)!
Thank you for your time and detailed review! I really felt my learn improving by seeing what was missing/ wrong while I solved all your comments. 🙏 

The main refactor of my review was separation of concerns, which I regret not to have cared before (time constrains).
Nevertheless it took me some time to move everything to their places, so I'm glad that I had the time to dive into the architecture of blackjack and figure out the concerns with time.

Let me know if any additional improvements are needed! 😄

Wordbe commented on src/main/kotlin/blackjack/GameLogic.kt:
```diff
@@ -0,0 +1,32 @@
+
+class GameLogic {
+    fun namesToPlayers(names: List<String>): List<Player> {
+         return names.map { name -> Player(name) }
```
Great 👍

Wordbe commented on src/main/kotlin/blackjack/Card.kt:
```diff
@@ -0,0 +1,19 @@
+        }
+    }
+
+    override fun toString(): String {
+        val cardValue = numberToCardDeckElements(number)
+        return "%s%s".format(cardValue, suit.value)
+    }
```
The domain object is currently handling code that's only needed in the View layer. Try to decouple the domain from the View by moving that logic into the View layer instead. 🙂

Wordbe reviewed (approved):
Hello!

Great job handling the reviews for the Blackjack mission.
It was nice to see how you abstracted `Participant` and extended it into `Player` and `Dealer`.

Please revisit the feedback I provided and try to apply it throughout the project as you go.
Make your code even more concise with Kotlin, aim for a clear separation between Model and View, and reinforce your tests overall.
Also, take a look at today’s review comments and reflect on what can be improved.

See you in step2. Great work! 👍
